packages/material-ui/src/Modal/TrapFocus.js

```diff
@@ -0,0 +1,146 @@
+/* eslint-disable consistent-return, jsx-a11y/no-noninteractive-tabindex */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import warning from 'warning';
+import RootRef from '../RootRef';
+import ownerDocument from '../utils/ownerDocument';
+
+function TrapFocus(props) {
+  const { disableEnforceFocus, disableAutoFocus, disableRestoreFocus, isEnabled, open } = props;
+  const rootRef = React.useRef();
+  const ignoreNextEnforceFocus = React.useRef();
+  const sentinelStart = React.useRef();
+  const sentinelEnd = React.useRef();
+  const lastFocus = React.useRef();
+
+  React.useEffect(() => {
+    if (!open) {
+      return;
+    }
+
+    const doc = ownerDocument(rootRef.current);
+    const currentActiveElement = doc.activeElement;
+    lastFocus.current = currentActiveElement;
+
+    // We might render an empty child.
+    if (!disableAutoFocus && rootRef.current && !rootRef.current.contains(currentActiveElement)) {
+      if (!rootRef.current.hasAttribute('tabIndex')) {
+        warning(
+          false,
+          [
+            'Material-UI: the modal content node does not accept focus.',
+            'For the benefit of assistive technologies, ' +
+              'the tabIndex of the node is being set to "-1".',
+          ].join('\n'),
+        );
+        rootRef.current.setAttribute('tabIndex', -1);
+      }
+
+      rootRef.current.focus();
+    }
+
+    const enforceFocus = () => {
+      if (disableEnforceFocus || !isEnabled() || ignoreNextEnforceFocus.current) {
+        ignoreNextEnforceFocus.current = false;
+        return;
+      }
+
+      if (!rootRef.current.contains(doc.activeElement)) {
+        rootRef.current.focus();
+      }
+    };
+
+    const loopFocus = event => {
```

Comment:
There is some option to avoid tabbing that `sentinelEnd` and `sentinelStart` elements ?

I noticed that if you have focus on the last element you need to press two times `TAB` key to go to the first element on the dialog and vice versa. In this case, screen reader vocalized nothing and user can be feel lost.

Reply:
Btw, what do you think on option to disable looping.

Reply:
I have tried to reproduce Bootstrap Modal's behavior: https://getbootstrap.com/docs/4.3/components/modal/. In order to implement what you are suggesting, we need a way to list all the dom elements in the modal that can receive the focus. It's not obvious, you would have to rely on something like this: [tabbable](https://bundlephobia.com/result?p=tabbable@3.1.2): +1kB gzipped.

(cc @ryancogswell, just for context)

Reply:
I've never tried, but it might work: 
- on `sentinel` focus
- list all nodes between `Start` and `End`
- try `.focus` each one
- if `.activeElement` changed - then node is focusable,
- break the loop.

Probably would not play well with radio buttons.

Reply:
Interesting :)